x/denommetadata/ibc_middleware.go

```diff
@@ -0,0 +1,244 @@
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+	"github.com/dymensionxyz/gerr-cosmos/gerrc"
+
+	utilsibc "github.com/dymensionxyz/dymension/v3/utils/ibc"
+	"github.com/dymensionxyz/dymension/v3/x/denommetadata/types"
+)
+
+var _ porttypes.IBCModule = &IBCModule{}
+
+// IBCModule implements the ICS26 callbacks for the transfer middleware
+type IBCModule struct {
+	porttypes.IBCModule
+	transferKeeper types.TransferKeeper
+	keeper         types.DenomMetadataKeeper
+	rollappKeeper  types.RollappKeeper
+}
+
+// NewIBCModule creates a new IBCModule given the keepers and underlying application
+func NewIBCModule(
+	app porttypes.IBCModule,
+	transferKeeper types.TransferKeeper,
+	keeper types.DenomMetadataKeeper,
+	rollappKeeper types.RollappKeeper,
+) IBCModule {
+	return IBCModule{
+		IBCModule:      app,
+		transferKeeper: transferKeeper,
+		keeper:         keeper,
+		rollappKeeper:  rollappKeeper,
+	}
+}
+
+// OnRecvPacket registers the denom metadata if it does not exist.
+// It will intercept an incoming packet and check if the denom metadata exists.
+// If it does not, it will register the denom metadata.
+// The handler will expect a 'denom_metadata' object in the memo field of the packet.
+// If the memo is not an object, or does not contain the metadata, it moves on to the next handler.
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	transferData, err := im.rollappKeeper.GetValidTransfer(ctx, packet.Data, packet.DestinationPort, packet.DestinationChannel)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	rollapp, packetData := transferData.Rollapp, transferData.FungibleTokenPacketData
+
+	if packetData.Memo == "" {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	// at this point it's safe to assume that we are not handling a native token of the rollapp
+	denomTrace := utilsibc.GetForeignDenomTrace(packet.GetDestChannel(), packetData.Denom)
+	ibcDenom := denomTrace.IBCDenom()
+
+	dm := types.ParsePacketMetadata(packetData.Memo)
+	if dm == nil {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	if err = dm.Validate(); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	// if denom metadata was found in the memo, it means we should have the rollapp record
+	if rollapp == nil {
+		return channeltypes.NewErrorAcknowledgement(gerrc.ErrNotFound)
+	}
+
+	dm.Base = ibcDenom
+	dm.DenomUnits[0].Denom = dm.Base
+
+	if err = im.keeper.CreateDenomMetadata(ctx, *dm); err != nil {
+		if errorsmod.IsOf(err, gerrc.ErrAlreadyExist) {
+			return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+		}
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if !im.transferKeeper.HasDenomTrace(ctx, denomTrace.Hash()) {
+		im.transferKeeper.SetDenomTrace(ctx, denomTrace)
```

Comment:
why u need to set it for the transferKeeper? isn't it gonna be created anyway in the inner `OnRecvPacket`?